### PR TITLE
feat!: Track class first iteration

### PR DESCRIPTION
**Pull request by paulgessinger**
Introduces a track class modeled after `MultiTrajectory`. In the current state, the KF, CKF, GSF and GX2F are changed to accept a `TrackContainer` as an argument. The KF, GSF and GX2F return a single `TrackProxy`, while the CKF returns a vector of `TrackProxy`s, that the calling code can use.

The caller has to recover certain information from the track, rather than having access to the internal result struct. I think this is fine as it makes the results much easier to use. OTOH some information is not accessible anymore.

I also illustrate the ability to add additional (debug) information with dynamic columns. If the track container has columns "reversed" and "smoothed", then the KF will add this information to the output track, allowing e.g. the KF tests to access this information. If the columns are not present, then it's not written. I expect / hope that this pattern also translates to e.g. the GSF.

The output tracks are in the current state immediately converted to `Trajectory` objects in the calling algorithms.
My plan would be to merge this in this form first, then transition to writing the track container out to the whiteboard directly, and adding an extra algorithm that can convert from tracks to trajectories. Finally, we would update the downstream performance writers etc. to operate on track container directly.

### CODE REVIEW

**Comment by paulgessinger** [comment]
Fyi @asalzburger @tboldagh @noemina

**Review by andiwand — COMMENTED** [review]
just went through quickly

File: Core/include/Acts/EventData/Track.hpp [hunk 1]
```diff
@@ -141,9 +143,9 @@ class TrackProxy {
   }
 
   template <bool RO = ReadOnly, typename = std::enable_if_t<!RO>>
-  void setReferenceSurface(std::shared_ptr<const Surface> srf) {
+  void setReferenceSurface(const std::shared_ptr<const Surface>& srf) {
```
[the remark this answers is not in the document]
Comment (andiwand) [reply Core/include/Acts/EventData/Track.hpp:146]: but this is what we saw here too no? https://github.com/acts-project/acts/blob/main/Core/include/Acts/EventData/MultiTrajectory.hpp#L425

should we use the same solution?

File: Alignment/include/ActsAlignment/Kernel/Alignment.hpp [hunk 2]
```diff
@@ -136,6 +136,8 @@ struct AlignmentResult {
 /// @tparam fitter_t Type of the fitter class
 template <typename fitter_t>
 struct Alignment {
+  // @TODO: Redefine in terms of Track object
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Alignment/include/ActsAlignment/Kernel/Alignment.hpp:139]: I thought it makes sense to have this as a reference. Can remove though if you prefer.

File: Core/include/Acts/EventData/Track.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/EventData/Track.hpp:146]: That's true, I'll change it.

File: Tests/UnitTests/Core/EventData/MultiTrajectoryTests.cpp [hunk 3]
```diff
@@ -233,6 +130,11 @@ BOOST_AUTO_TEST_CASE(AsReadOnly) {
     static_cast<void>(tsp);
     VectorMultiTrajectory::ConstTrackStateProxy ctsp = t.getTrackState(i0);
     static_cast<void>(ctsp);
+
+    tsp.predicted().setRandom();
+    // const auto& tsp_const = tsp;
+    // tsp_const.predicted().setRandom();
+    // ctsp.predicted().setRandom();
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Tests/UnitTests/Core/EventData/MultiTrajectoryTests.cpp:137]: This shouldn't compile, so I leave it commented out so we can test it manually if needed.

File: Core/include/Acts/EventData/VectorTrackContainer.hpp [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+#include "Acts/EventData/Track.hpp"
+
+#include <unordered_map>
+
+namespace Acts {
+
+namespace detail_vtc {
+
+struct DynamicColumnBase {
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/EventData/VectorTrackContainer.hpp:19]: Yeah, I wasn't sure if I needed changes here, but I guess I could reuse the one in MTJ.

File: Examples/Algorithms/TrackFitting/src/TrackFittingAlgorithm.cpp [hunk 5]
```diff
@@ -133,29 +136,32 @@ ActsExamples::ProcessCode ActsExamples::TrackFittingAlgorithm::execute(
...
       std::vector<Acts::MultiTrajectoryTraits::IndexType> trackTips;
       trackTips.reserve(1);
-      trackTips.emplace_back(fitOutput.lastMeasurementIndex);
+      trackTips.emplace_back(track.tipIndex());
       // The fitted parameters container. One element (at most) here.
       Trajectories::IndexedParameters indexedParams;
-      if (fitOutput.fittedParameters) {
-        const auto& params = fitOutput.fittedParameters.value();
+      if (track.hasReferenceSurface()) {
+        // const auto& params = fitOutput.fittedParameters.value();
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Examples/Algorithms/TrackFitting/src/TrackFittingAlgorithm.cpp:152]:
```suggestion
```

File: Alignment/include/ActsAlignment/Kernel/Alignment.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (andiwand) [reply Alignment/include/ActsAlignment/Kernel/Alignment.hpp:139]: oh yes this makes sense as it is the followup work

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 6]
```diff
@@ -212,7 +214,7 @@ struct CombinatorialKalmanFilterOptions {
 template <typename traj_t>
 struct CombinatorialKalmanFilterResult {
   // Fitted states that the actor has handled.
-  std::shared_ptr<traj_t> fittedStates;
+  traj_t* fittedStates{nullptr};
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:217]: It's not but I'm not making the assumption that the track container owns the MTJ. So I'm changing this again to only keep a non owning pointer in the fitters. The background is that I want to allow standalone to have the track container own everything, but in atlas to have store gate own the container backends.

File: Core/include/Acts/EventData/VectorTrackContainer.hpp [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+        *this, key, itrack);
+  }
+
+  std::any component_impl(HashedString key, IndexType itrack) const {
+    return detail_vtc::VectorTrackContainerBase::component_impl<true>(
+        *this, key, itrack);
+  }
+
+  IndexType addTrack_impl() {
+    m_tipIndex.emplace_back();
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/EventData/VectorTrackContainer.hpp:206]: It's the entry point. The link to the MTJ is an explicit pointer, or value that this links into. 

Are you thinking about a use case where the track doesn't have track states? 

The point about ROIs is interesting. My intention was indeed to have one track collection refer to a single MTJ. I didn't fully consider this; we expect the ROIs to feed into a single track container in the end? Is this currently synchronized between threads or just one after the other in each event?

File: Core/include/Acts/EventData/MultiTrajectory.hpp [hunk 8]
```diff
@@ -271,7 +273,7 @@ class TrackStateProxy {
   ///       not allocated in the source track state proxy.
   template <bool RO = ReadOnly, bool ReadOnlyOther,
             typename = std::enable_if_t<!RO>>
-  void copyFrom(const TrackStateProxy<Trajectory, M, ReadOnlyOther>& other,
+  void copyFrom(TrackStateProxy<Trajectory, M, ReadOnlyOther> other,
```
[the remark this answers is not in the document]
Comment (tboldagh) [reply Core/include/Acts/EventData/MultiTrajectory.hpp:276]: Ah I somehow misremembered that it is a bigger class.

File: Core/include/Acts/EventData/VectorMultiTrajectory.hpp [hunk 9]
```diff
@@ -24,6 +24,72 @@ using MultiTrajectoryTraits::IndexType;
 constexpr auto kInvalid = MultiTrajectoryTraits::kInvalid;
 constexpr auto MeasurementSizeMax = MultiTrajectoryTraits::MeasurementSizeMax;
 
+struct DynamicColumnBase {
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/EventData/VectorMultiTrajectory.hpp]: I can pull this out I guess, yeah

File: Core/include/Acts/EventData/Track.hpp [hunk 10]
```diff
@@ -0,0 +1,573 @@
...
+#pragma once
+
+#include "Acts/EventData/MultiTrajectory.hpp"
+#include "Acts/Utilities/HashedString.hpp"
+
+#include <any>
+#include <cstddef>
+#include <iterator>
+
+#include <unistd.h>
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/EventData/Track.hpp:18]: Vim added this for me. I'll try without.

File: Core/include/Acts/EventData/Track.hpp [hunk 11]
```diff
@@ -0,0 +1,573 @@
...
+      : m_container{&container}, m_index{itrack} {}
+
+  detail_lt::TransitiveConstPointer<
+      ConstIf<TrackContainer<Container, Trajectory, holder_t>, ReadOnly>>
+      m_container;
+  IndexType m_index;
+};
+
+template <typename T>
+struct RefHolder {
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/EventData/Track.hpp:232]: Seems like an implementation detail. I didn't want to make this a "thing". I think I'd prefer to keep it this way until we use it somewhere else I think.

File: Core/include/Acts/EventData/VectorMultiTrajectory.hpp [hunk 12]
```diff
@@ -24,6 +24,72 @@ using MultiTrajectoryTraits::IndexType;
...
+  virtual void clear() = 0;
+
+  virtual std::unique_ptr<DynamicColumnBase> clone() const = 0;
+};
+
+inline DynamicColumnBase::~DynamicColumnBase() = default;
+
+template <typename T>
+struct DynamicColumn : public DynamicColumnBase {
+  ~DynamicColumn() override = default;
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/EventData/VectorMultiTrajectory.hpp]: Don't I need it I make the base class destructor pure virtual? And I do want a virtual destructor so it deallocates the vectors.

File: Core/include/Acts/EventData/VectorTrackContainer.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (tboldagh) [reply Core/include/Acts/EventData/VectorTrackContainer.hpp:206]: Concerning first. My point is that in final analysis format the MTJ will be absent. So this index to entry point will be kind of dangling reference. I guess we arrange it like this in ATLAS and do not have to in ACTS.

At the moment (and likely in the future) the arrangement one-RoI -> one-MTJ -> one-tracks collection is reasonable. Many RoIs reconstructed in parallel potentially.
We then merge thole collection for persistency but this is detail as it concerns small fraction of passing events.  But I still wonder how we will do the bookkeeping.

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (tboldagh) [reply Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:217]: ok, sure

File: Core/include/Acts/EventData/VectorMultiTrajectory.hpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (andiwand) [reply Core/include/Acts/EventData/VectorMultiTrajectory.hpp]: the base class for sure but I thought the derived class destructor is implicitly defaulted

File: Core/include/Acts/EventData/Track.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/EventData/Track.hpp:146]: ✅

File: Core/include/Acts/EventData/Track.hpp [hunk 13]
```diff
@@ -0,0 +1,573 @@
...
+ private:
+  detail_lt::TransitiveConstPointer<ConstIf<ContainerType, ReadOnly>>
+      m_container;
+  IndexType m_itrack;
+};
+
+}  // namespace detail_tc
+
+template <typename T>
+struct isReadOnlyTrackContainer;
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/EventData/Track.hpp:378]: ✅

File: Core/include/Acts/EventData/VectorMultiTrajectory.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/EventData/VectorMultiTrajectory.hpp]: Done now.

File: Core/include/Acts/EventData/VectorTrackContainer.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (tboldagh) [reply Core/include/Acts/EventData/VectorTrackContainer.hpp:141]: Indeed. All good.

**Review by tboldagh — APPROVED** [review]

File: Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/TrackFindingAlgorithm.hpp [hunk 14]
```diff
@@ -125,56 +130,41 @@ void TrackFindingAlgorithm::computeSharedHits(
...
-            .fittedStates->getTrackState(state.index())
-            .typeFlags()
-            .set(Acts::TrackStateFlag::SharedHitFlag);
-
-        return true;
-      });
+  for (auto track : tracks) {
+    for (auto state : track.trackStates()) {
+      if (not state.typeFlags().test(Acts::TrackStateFlag::MeasurementFlag)) {
+        break;
```
Comment (paulgessinger) [inline Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/TrackFindingAlgorithm.hpp:136]:
```suggestion
        continue;
```

File: Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/TrackFindingAlgorithm.hpp [hunk 15]
```diff
@@ -125,56 +130,41 @@ void TrackFindingAlgorithm::computeSharedHits(
...
+
+      std::size_t hitIndex =
+          state.uncalibrated().template get<IndexSourceLink>().index();
+
+      // Check if hit not already used
+      if (firstTrackOnTheHit.at(hitIndex) ==
+          std::numeric_limits<std::size_t>::max()) {
+        firstTrackOnTheHit.at(hitIndex) = track.index();
+        firstStateOnTheHit.at(hitIndex) = state.index();
+        break;
```
Comment (paulgessinger) [inline Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/TrackFindingAlgorithm.hpp:147]:
```suggestion
        continue;
```

File: Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/TrackFindingAlgorithm.hpp [hunk 16]
```diff
@@ -125,56 +130,41 @@ void TrackFindingAlgorithm::computeSharedHits(
...
+                            .getTrackState(indexFirstState);
+      if (not firstState.typeFlags().test(
+              Acts::TrackStateFlag::SharedHitFlag)) {
+        firstState.typeFlags().set(Acts::TrackStateFlag::SharedHitFlag);
+      }
+
+      // Decorate this track
+      state.typeFlags().set(Acts::TrackStateFlag::SharedHitFlag);
+
+      break;
```
Comment (paulgessinger) [inline Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/TrackFindingAlgorithm.hpp:167]:
```suggestion
      continue;
```
or remove completely.

File: Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/TrackFindingAlgorithm.hpp [hunk 17]
```diff
@@ -125,56 +130,41 @@ void TrackFindingAlgorithm::computeSharedHits(
...
+      auto firstState = tracks.getTrack(indexFirstTrack)
+                            .container()
+                            .trackStateContainer()
+                            .getTrackState(indexFirstState);
+      if (not firstState.typeFlags().test(
+              Acts::TrackStateFlag::SharedHitFlag)) {
+        firstState.typeFlags().set(Acts::TrackStateFlag::SharedHitFlag);
+      }
+
+      // Decorate this track
```
Comment (paulgessinger) [inline Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/TrackFindingAlgorithm.hpp:164]:
```suggestion
      // Decorate this track state
```

**Comment by paulgessinger** [comment]
This is ready to merge I guess. @andiwand @tboldagh @asalzburger

**Comment by tboldagh** [comment]
> This is ready to merge I guess. 
Definitely so. I started to look at the ATLAS backend.

**Comment by paulgessinger** [comment]
Could you reapprove so I can merge?

**Review by tboldagh — APPROVED** [review]

**Pull request merged**